include/fmt/core.h

```diff
@@ -457,15 +457,23 @@ template <typename Char> class basic_string_view {
   }
 
   // Lexicographically compare this string reference to other.
-  int compare(basic_string_view other) const {
+#if __cplusplus >= 201703L  // C++17's char_traits::compare() is constexpr.
+  constexpr
+#endif
```

Comment:
Using the ``__cplusplus`` macro to determine the available features of the standard library may not work.

For example, if the system compiler is gcc-6, and you are using new clang 9 with ``-std=gnu++17 -stdlib=libstdc++``, then checking the ``__cplusplus`` value will give an incorrect result (libstdc++ from gcc 6 doesn't have constexpr ``std::char_traits<>::compare``).

Reply:
You are definitely right here.
In my defense, I can say that I'm not the first who did this 🙂:
https://github.com/fmtlib/fmt/blob/52bd62c72f6c15dcfaee237df09cdafd9a990fc3/include/fmt/core.h#L418
Also, I didn't find the corresponding feature here - https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros, but looks like it's presented in SD-FeatureTest: https://isocpp.org/std/standing-documents/sd-6-sg10-feature-test-recommendations#__cpp_lib_constexpr_string

Reply:
But a [quick test](https://godbolt.org/z/zzvfhh5hq) for this macro shows that it's not defined like that on both GCC and Clang.
So I don't know how to check the availability of this feature then.

Reply:
``__cpp_lib_constexpr_string`` - c++20.

This problem only on libstdc++.

Working check (for constexpr char_traits):
``__cplusplus >= 201703L && (!defined(__GLIBCXX__) || (defined(__GLIBCXX__) && defined(_GLIBCXX_RELEASE)))``

``constexpr std::char_traits<>::compare`` is present start from gcc 7.

The macro _GLIBCXX_RELEASE is first defined in GCC 7.1 release.

Reply:
> `__cpp_lib_constexpr_string` - c++20.

As described in [feature test recommendations document](https://isocpp.org/std/standing-documents/sd-6-sg10-feature-test-recommendations#__cpp_lib_constexpr_string), it's not C++20, it should be defined to `201611` if _[P0426R1] Constexpr for std::char_traits_ is implemented. But looks like GCC and LLVM folks didn't follow these recommendations.

I think it's probably out of the scope of this PR since this problem is also presented in other place(s). So if you would implement this check correctly then I can just rebase and use this check here 😉

Reply:
PR: https://github.com/fmtlib/fmt/pull/2246

Reply:
Let's at least get rid of copy-paste in this PR and introduce `FMT_CONSTEXPR_STRING` macro that expands to `constexpr` if `char_traits`/`string` operations are `constexpr`.

Reply:
Maybe ``FMT_CONSTEXPR_CHAR_TRAITS``, because ``std::basic_string`` operations is ``constexpr`` only from C++20?

Reply:
Sure

Reply:
PR https://github.com/fmtlib/fmt/pull/2246 is updated.

Reply:
#2246 has been merged, please rebase.

Reply:
Rebased